lua/ttt2/libraries/events.lua

```diff
@@ -0,0 +1,204 @@
+	for k, v in pairs(base) do
+		if t[k] == nil then
+			t[k] = v
+		elseif k ~= "BaseClass" and istable(t[k]) then
+			TableInherit(t[k], v)
+		end
+	end
+
+	t.BaseClass = base
+
+	return t
+end
+
+---
+-- This function is called once an event is loaded from disc.
+-- @param string path The path to the event file
+-- @internal
+-- @realm shared
+function events.Initialize(path)
+	local pathArray = string.Split(path, "/")
+	local name = string.lower(string.sub(pathArray[#pathArray], 0, -5))
+
+	-- event table is set in fileloader and can now be inserted in the table
+	local newEvent = tableCopy(EVENT)
+	newEvent.type = name
+	newEvent.base = newEvent.base or "base_event"
+
+	eventTypes[name] = newEvent
+
+	-- store a global identifier for this event
+	_G["EVENT_" .. string.upper(name)] = name
+
+	-- reset EVENT table
+	EVENT = {}
+end
+
+---
+-- Initialialized the events after everything is loaded. This included the
+-- inheritance from their base classes.
+-- @internal
+-- @realm shared
+function events.OnLoaded()
+	for index, event in pairs(eventTypes) do
+		eventTypes[index] = TableInherit(event, events.Get(event.base))
+	end
+end
+
+---
+-- Returns a table of all available events.
+-- @return table A table of all events
+-- @realm shared
+function events.GetAll()
+	return eventTypes
+end
+
+---
+-- Returns the reference to a table of the event.
+-- @param string name The name identifer of the event
+-- @return table The reference to the event table
+-- @realm shared
+function events.Get(name)
+	return eventTypes[string.lower(name)]
+end
+
+---
+-- Returns the reference to a copy of the table of the event.
+-- @param string name The name identifer of the event
+-- @return table The reference to the copied event table
+-- @realm shared
+function events.New(name)
+	return tableCopy(events.Get(name))
+end
+
+---
+-- Checks if an event with this ID is registered in the event list.
+-- @param string name The name identifer of the event
+-- @return boolean Returns true if an event of this name exists
+-- @realm shared
+function events.Exist(name)
+	return events.Get(name) ~= nil
+end
+
+---
+-- Resets the managed event list. Usually executed after the end of the round end timer.
+-- @internal
+-- @realm shared
+function events.Reset()
+	events.list = {}
+end
+
+if SERVER then
+	---
+	-- Triggers an event, adds it to a managed list and starts the score calculation for this event.
+	-- @param string name The name identifer of the event
+	-- @param varag The arguments that should be passed to the event, see the @{EVENT:Trigger} function
+	-- @realm server
+	function events.Trigger(name, ...)
+		if not events.Exist(name) then
+			ErrorNoHalt("ERROR: An event with the name '" .. tostring(name) .. "' does not exist.\n")
+
+			return
+		end
+
+		local newEvent = events.New(name)
+		local eventData = newEvent:Trigger(...)
+
+		-- only add new event to managed event list, if addition was not aborted
+		if newEvent:Add(eventData) then
```

Comment:
I know why you doing this, but it's a bit weird, because at first you calling a method on the new event to create the eventData and after that, you adding the created eventData to the newEvent again. We already talked about it (the current and previous solution) and I definitely prefer the way it's done NOW (just returning the event table in the Trigger method is still the best solution). But maybe someone else has a better solution...